File: defence_claw/defenseclaw/internal/enforce/policy.go
// Copyright 2026 Cisco Systems, Inc. and its affiliates
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package enforce

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"

	"github.com/defenseclaw/defenseclaw/internal/audit"
)

type PolicyEngine struct {
	store *audit.Store
}

func NewPolicyEngine(store *audit.Store) *PolicyEngine {
	return &PolicyEngine{store: store}
}

func (e *PolicyEngine) IsBlocked(targetType, name string) (bool, error) {
	if e.store == nil {
		return false, nil
	}
	return e.store.HasAction(targetType, name, "install", "block")
}

func (e *PolicyEngine) IsAllowed(targetType, name string) (bool, error) {
	if e.store == nil {
		return false, nil
	}
	return e.store.HasAction(targetType, name, "install", "allow")
}

func (e *PolicyEngine) IsQuarantined(targetType, name string) (bool, error) {
	if e.store == nil {
		return false, nil
	}
	return e.store.HasAction(targetType, name, "file", "quarantine")
}

func (e *PolicyEngine) Block(targetType, name, reason string) error {
	if e.store == nil {
		return nil
	}
	return e.store.SetActionField(targetType, name, "install", "block", reason)
}

func (e *PolicyEngine) Allow(targetType, name, reason string) error {
	if e.store == nil {
		return nil
	}
	if err := e.store.SetActionField(targetType, name, "install", "allow", reason); err != nil {
		return err
	}
	// Clear residual auto-enforcement state (quarantine / disable) so the
	// allow actually takes full effect.  Only a manual Block can override.
	var errs []error
	if err := e.store.ClearActionField(targetType, name, "file"); err != nil {
		errs = append(errs, fmt.Errorf("clear file action: %w", err))
	}
	if err := e.store.ClearActionField(targetType, name, "runtime"); err != nil {
		errs = append(errs, fmt.Errorf("clear runtime action: %w", err))
	}
	if len(errs) > 0 {
		return fmt.Errorf("enforce: allow %s %q: partial cleanup: %v", targetType, name, errs)
	}
	return nil
}

func (e *PolicyEngine) Unblock(targetType, name string) error {
	if e.store == nil {
		return nil
	}
	return e.store.ClearActionField(targetType, name, "install")
}

func (e *PolicyEngine) Quarantine(targetType, name, reason string) error {
	if e.store == nil {
		return nil
	}
	return e.store.SetActionField(targetType, name, "file", "quarantine", reason)
}

func (e *PolicyEngine) ClearQuarantine(targetType, name string) error {
	if e.store == nil {
		return nil
	}
	return e.store.ClearActionField(targetType, name, "file")
}

func (e *PolicyEngine) Disable(targetType, name, reason string) error {
	if e.store == nil {
		return nil
	}
	return e.store.SetActionField(targetType, name, "runtime", "disable", reason)
}

func (e *PolicyEngine) Enable(targetType, name string) error {
	if e.store == nil {
		return nil
	}
	return e.store.ClearActionField(targetType, name, "runtime")
}

func (e *PolicyEngine) SetSourcePath(targetType, name, path string) {
	if e.store == nil {
		return
	}
	_ = e.store.SetSourcePath(targetType, name, path)
}

func (e *PolicyEngine) SetAction(targetType, name, sourcePath string, state audit.ActionState, reason string) error {
	if e.store == nil {
		return nil
	}
	return e.store.SetAction(targetType, name, sourcePath, state, reason)
}

func (e *PolicyEngine) GetAction(targetType, name string) (*audit.ActionEntry, error) {
	if e.store == nil {
		return nil, nil
	}
	return e.store.GetAction(targetType, name)
}

func (e *PolicyEngine) ListBlocked() ([]audit.ActionEntry, error) {
	if e.store == nil {
		return nil, nil
	}
	return e.store.ListByAction("install", "block")
}

func (e *PolicyEngine) ListAllowed() ([]audit.ActionEntry, error) {
	if e.store == nil {
		return nil, nil
	}
	return e.store.ListByAction("install", "allow")
}

func (e *PolicyEngine) ListAll() ([]audit.ActionEntry, error) {
	if e.store == nil {
		return nil, nil
	}
	return e.store.ListAllActions()
}

func (e *PolicyEngine) ListByType(targetType string) ([]audit.ActionEntry, error) {
	if e.store == nil {
		return nil, nil
	}
	return e.store.ListActionsByType(targetType)
}

func (e *PolicyEngine) RemoveAction(targetType, name string) error {
	if e.store == nil {
		return nil
	}
	return e.store.RemoveAction(targetType, name)
}

// ----------------------------------------------------------------------------
// Connector-scoped enforcement helpers (N2 — per-connector mcp
// block/allow/unblock)
//
// The connector dimension lives in the audit store's per-connector "connector"
// column (the SK-4 foundation), which is distinct from the
// "@<connector>/<tool>" name-encoding the tool gate uses below. A bare entry
// (connector="") is GLOBAL — it applies to every connector; a non-empty
// connector NARROWS the entry to that peer.
//
// Reads resolve most-specific-wins: a connector-scoped install action decides
// for that connector, then the global entry falls through only when no scoped
// install action exists. That means a global block still applies broadly, while
// a connector-scoped allow can override it for that connector. Writes are
// exact-match on connector (the actions table is unique on
// (target_type, target_name, connector)). Mirrors the *_for_connector methods
// in cli/defenseclaw/enforce/policy.py.
// ----------------------------------------------------------------------------

// IsBlockedForConnector reports whether name is blocked for connector, checking
// the connector-scoped entry first and then the bare global entry.
func (e *PolicyEngine) IsBlockedForConnector(targetType, name, connector string) (bool, error) {
	if e.store == nil {
		return false, nil
	}
	if connector != "" {
		action, ok, err := e.actionFieldForConnector(targetType, name, connector, "install")
		if err != nil {
			return false, err
		}
		if ok {
			return action == "block", nil
		}
	}
	return e.store.HasAction(targetType, name, "install", "block")
}

// IsAllowedForConnector reports whether name is allowed for connector, checking
// the connector-scoped entry first and then the bare global entry. A
// connector-scoped block is authoritative for that connector, so callers that
// check IsBlockedForConnector first will still block it before consulting allow.
func (e *PolicyEngine) IsAllowedForConnector(targetType, name, connector string) (bool, error) {
	if e.store == nil {
		return false, nil
	}
	if connector != "" {
		action, ok, err := e.actionFieldForConnector(targetType, name, connector, "install")
		if err != nil {
			return false, err
		}
		if ok {
			return action == "allow", nil
		}
	}
	return e.store.HasAction(targetType, name, "install", "allow")
}

// IsQuarantinedForConnector reports whether name is quarantined for connector,
// checking the connector-scoped entry first and then the bare global entry.
func (e *PolicyEngine) IsQuarantinedForConnector(targetType, name, connector string) (bool, error) {
	if e.store == nil {
		return false, nil
	}
	if connector != "" {
		action, ok, err := e.actionFieldForConnector(targetType, name, connector, "file")
		if err != nil {
			return false, err
		}
		if ok {
			return action == "quarantine", nil
		}
	}
	return e.store.HasAction(targetType, name, "file", "quarantine")
}

// IsDisabledForConnector reports whether name is runtime-disabled for
// connector, checking the connector-scoped entry first and then the bare global
// entry.
func (e *PolicyEngine) IsDisabledForConnector(targetType, name, connector string) (bool, error) {
	if e.store == nil {
		return false, nil
	}
	if connector != "" {
		action, ok, err := e.actionFieldForConnector(targetType, name, connector, "runtime")
		if err != nil {
			return false, err
		}
		if ok {
			return action == "disable", nil
		}
	}
	return e.store.HasAction(targetType, name, "runtime", "disable")
}

func (e *PolicyEngine) actionFieldForConnector(targetType, name, connector, field string) (string, bool, error) {
	entry, err := e.store.GetActionForConnector(targetType, name, connector)
	if err != nil {
		return "", false, err
	}
	if entry == nil {
		return "", false, nil
	}
	return actionFieldValue(entry.Actions, field)
}

func actionFieldValue(actions audit.ActionState, field string) (string, bool, error) {
	switch field {
	case "install":
		return actions.Install, actions.Install != "", nil
	case "file":
		return actions.File, actions.File != "", nil
	case "runtime":
		return actions.Runtime, actions.Runtime != "", nil
	default:
		return "", false, fmt.Errorf("enforce: unknown action field %q", field)
	}
}

// BlockForConnector blocks name for connector (exact-match; connector="" = global).
func (e *PolicyEngine) BlockForConnector(targetType, name, connector, reason string) error {
	if e.store == nil {
		return nil
	}
	return e.store.SetActionFieldForConnector(targetType, name, connector, "install", "block", reason)
}

// AllowForConnector allows name for connector and clears residual file/runtime
// enforcement (exact-match; connector="" = global). Mirrors Allow().
func (e *PolicyEngine) AllowForConnector(targetType, name, connector, reason string) error {
	if e.store == nil {
		return nil
	}
	if err := e.store.SetActionFieldForConnector(targetType, name, connector, "install", "allow", reason); err != nil {
		return err
	}
	var errs []error
	if err := e.store.ClearActionFieldForConnector(targetType, name, connector, "file"); err != nil {
		errs = append(errs, fmt.Errorf("clear file action: %w", err))
	}
	if err := e.store.ClearActionFieldForConnector(targetType, name, connector, "runtime"); err != nil {
		errs = append(errs, fmt.Errorf("clear runtime action: %w", err))
	}
	if len(errs) > 0 {
		return fmt.Errorf("enforce: allow %s %q connector %q: partial cleanup: %v", targetType, name, connector, errs)
	}
	return nil
}

// UnblockForConnector clears the install action for connector (exact-match;
// connector="" = global).
func (e *PolicyEngine) UnblockForConnector(targetType, name, connector string) error {
	if e.store == nil {
		return nil
	}
	return e.store.ClearActionFieldForConnector(targetType, name, connector, "install")
}

// RemoveActionForConnector removes all enforcement for connector (exact-match;
// connector="" = global).
func (e *PolicyEngine) RemoveActionForConnector(targetType, name, connector string) error {
	if e.store == nil {
		return nil
	}
	return e.store.RemoveActionForConnector(targetType, name, connector)
}

// ----------------------------------------------------------------------------
// Connector-scoped tool helpers (targetType="tool", key "@<connector>/<tool>")
//
// The "@" sigil keeps connector scoping distinct from the orthogonal
// "<source>/<tool>" source scoping used by the CLI. Runtime resolution order,
// for request connector C and tool T, is:
//
//	action @C/T → action T → scan
//
// i.e. the connector-scoped install action decides for that connector, then the
// bare global install action is used only when there is no scoped install
// action. Mirrors the Python PolicyEngine methods of the same name in
// cli/defenseclaw/enforce/policy.py.
// ----------------------------------------------------------------------------

// toolConnectorTarget builds the connector-scoped tool key "@<connector>/<tool>".
// Centralised so the read gate and any write surface stay in lockstep.
func toolConnectorTarget(toolName, connector string) string {
	if connector == "" {
		return toolName
	}
	return "@" + connector + "/" + toolName
}

// IsToolBlockedForConnector reports whether toolName is blocked for connector,
// checking the connector-scoped entry first and then the bare global entry.
func (e *PolicyEngine) IsToolBlockedForConnector(toolName, connector string) (bool, error) {
	if e.store == nil {
		return false, nil
	}
	if connector != "" {
		scoped := toolConnectorTarget(toolName, connector)
		action, ok, err := e.actionField("tool", scoped, "install")
		if err != nil {
			return false, err
		}
		if ok {
			return action == "block", nil
		}
	}
	return e.store.HasAction("tool", toolName, "install", "block")
}

// IsToolAllowedForConnector reports whether toolName is allowed for connector,
// checking the connector-scoped entry first and then the bare global entry. A
// connector-scoped block is authoritative for that connector, so callers that
// check IsToolBlockedForConnector first will still block it before consulting
// allow.
func (e *PolicyEngine) IsToolAllowedForConnector(toolName, connector string) (bool, error) {
	if e.store == nil {
		return false, nil
	}
	if connector != "" {
		scoped := toolConnectorTarget(toolName, connector)
		action, ok, err := e.actionField("tool", scoped, "install")
		if err != nil {
			return false, err
		}
		if ok {
			return action == "allow", nil
		}
	}
	return e.store.HasAction("tool", toolName, "install", "allow")
}

func (e *PolicyEngine) actionField(targetType, name, field string) (string, bool, error) {
	entry, err := e.store.GetAction(targetType, name)
	if err != nil {
		return "", false, err
	}
	if entry == nil {
		return "", false, nil
	}
	return actionFieldValue(entry.Actions, field)
}

// BlockToolForConnector blocks toolName, optionally scoped to a connector.
func (e *PolicyEngine) BlockToolForConnector(toolName, connector, reason string) error {
	if e.store == nil {
		return nil
	}
	return e.store.SetActionField("tool", toolConnectorTarget(toolName, connector), "install", "block", reason)
}

// AllowToolForConnector allows toolName, optionally scoped to a connector, and
// clears residual file/runtime enforcement (mirrors Allow()).
func (e *PolicyEngine) AllowToolForConnector(toolName, connector, reason string) error {
	if e.store == nil {
		return nil
	}
	target := toolConnectorTarget(toolName, connector)
	if err := e.store.SetActionField("tool", target, "install", "allow", reason); err != nil {
		return err
	}
	var errs []error
	if err := e.store.ClearActionField("tool", target, "file"); err != nil {
		errs = append(errs, fmt.Errorf("clear file action: %w", err))
	}
	if err := e.store.ClearActionField("tool", target, "runtime"); err != nil {
		errs = append(errs, fmt.Errorf("clear runtime action: %w", err))
	}
	if len(errs) > 0 {
		return fmt.Errorf("enforce: allow tool %q: partial cleanup: %v", target, errs)
	}
	return nil
}

// PolicyStableID returns a short stable identifier for the policy bundle
// rooted at policyDir (used in OTel spans and metrics).
func PolicyStableID(policyDir string) string {
	if policyDir == "" {
		return "none"
	}
	sum := sha256.Sum256([]byte(policyDir))
	return hex.EncodeToString(sum[:8])
}

File: defence_claw/defenseclaw/cli/defenseclaw/enforce/policy.py
# Copyright 2026 Cisco Systems, Inc. and its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""PolicyEngine — thin facade over the audit Store for enforcement decisions.

Mirrors internal/enforce/policy.go exactly.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from defenseclaw.models import ActionEntry, ActionState

if TYPE_CHECKING:
    from defenseclaw.db import Store


class PolicyEngine:
    def __init__(self, store: Store | None) -> None:
        self.store = store

    def is_blocked(self, target_type: str, name: str) -> bool:
        if not self.store:
            return False
        return self.store.has_action(target_type, name, "install", "block")

    def is_allowed(self, target_type: str, name: str) -> bool:
        if not self.store:
            return False
        return self.store.has_action(target_type, name, "install", "allow")

    def is_quarantined(self, target_type: str, name: str) -> bool:
        if not self.store:
            return False
        return self.store.has_action(target_type, name, "file", "quarantine")

    def block(self, target_type: str, name: str, reason: str) -> None:
        if self.store:
            self.store.set_action_field(target_type, name, "install", "block", reason)

    def allow(self, target_type: str, name: str, reason: str) -> None:
        """Set install=allow and clear residual file/runtime enforcement.

        Mirrors internal/enforce/policy.go Allow() exactly: after allowing,
        quarantine and disable state are removed so the allow takes full
        effect.  Only a manual block() can override an allow entry.
        """
        if not self.store:
            return
        self.store.set_action_field(target_type, name, "install", "allow", reason)
        self.store.clear_action_field(target_type, name, "file")
        self.store.clear_action_field(target_type, name, "runtime")

    def unblock(self, target_type: str, name: str) -> None:
        if self.store:
            self.store.clear_action_field(target_type, name, "install")

    def quarantine(self, target_type: str, name: str, reason: str) -> None:
        if self.store:
            self.store.set_action_field(target_type, name, "file", "quarantine", reason)

    def clear_quarantine(self, target_type: str, name: str) -> None:
        if self.store:
            self.store.clear_action_field(target_type, name, "file")

    def disable(self, target_type: str, name: str, reason: str) -> None:
        if self.store:
            self.store.set_action_field(target_type, name, "runtime", "disable", reason)

    def enable(self, target_type: str, name: str) -> None:
        if self.store:
            self.store.clear_action_field(target_type, name, "runtime")

    def set_source_path(
        self, target_type: str, name: str, path: str, connector: str = "",
    ) -> None:
        if self.store:
            self.store.set_source_path(target_type, name, path, connector)

    def set_action(
        self, target_type: str, name: str, source_path: str,
        state: ActionState, reason: str,
    ) -> None:
        if self.store:
            self.store.set_action(target_type, name, source_path, state, reason)

    def get_action(
        self, target_type: str, name: str, connector: str = "",
    ) -> ActionEntry | None:
        if not self.store:
            return None
        return self.store.get_action(target_type, name, connector)

    def list_blocked(self) -> list[ActionEntry]:
        if not self.store:
            return []
        return self.store.list_by_action("install", "block")

    def list_allowed(self) -> list[ActionEntry]:
        if not self.store:
            return []
        return self.store.list_by_action("install", "allow")

    def list_all(self) -> list[ActionEntry]:
        if not self.store:
            return []
        return self.store.list_all_actions()

    def list_by_type(self, target_type: str) -> list[ActionEntry]:
        if not self.store:
            return []
        return self.store.list_actions_by_type(target_type)

    def remove_action(self, target_type: str, name: str) -> None:
        if self.store:
            self.store.remove_action(target_type, name)

    # ------------------------------------------------------------------
    # Connector-scoped enforcement helpers (N2 — per-connector
    # mcp block/allow/unblock)
    #
    # The connector dimension lives in the audit store's per-connector
    # ``connector`` column (the f/dbmig SK-4 foundation), which is distinct
    # from the ``@<connector>/<tool>`` name-encoding the tool gate uses below.
    # A bare entry (connector="") is **GLOBAL** — it applies to every
    # connector; a non-empty connector **NARROWS** the entry to that peer.
    #
    # Reads resolve **most-specific-wins per action field**: if the connector
    # owns a row with the requested field set, that field is authoritative for
    # that connector; otherwise the global row falls through. This lets a
    # connector-scoped allow override a global block for that connector, while a
    # connector-scoped block still wins when both scoped/global allows exist.
    # Writes are exact-match on connector (the actions table is unique on
    # (target_type, target_name, connector)). Mirrors the ``*ForConnector``
    # methods in internal/enforce/policy.go.
    # ------------------------------------------------------------------

    def is_blocked_for_connector(
        self, target_type: str, name: str, connector: str = "",
    ) -> bool:
        """True if blocked for ``connector`` (connector-scoped entry, else global)."""
        if not self.store:
            return False
        if connector:
            scoped = self.store.get_action(target_type, name, connector)
            if scoped is not None and scoped.actions.install:
                return scoped.actions.install == "block"
        return self.store.has_action(target_type, name, "install", "block")

    def is_allowed_for_connector(
        self, target_type: str, name: str, connector: str = "",
    ) -> bool:
        """True if allowed for ``connector`` (connector-scoped entry, else global)."""
        if not self.store:
            return False
        if connector:
            scoped = self.store.get_action(target_type, name, connector)
            if scoped is not None and scoped.actions.install:
                return scoped.actions.install == "allow"
        return self.store.has_action(target_type, name, "install", "allow")

    def is_quarantined_for_connector(
        self, target_type: str, name: str, connector: str = "",
    ) -> bool:
        """True if quarantined for ``connector`` (connector-scoped entry, else global)."""
        if not self.store:
            return False
        if connector:
            scoped = self.store.get_action(target_type, name, connector)
            if scoped is not None and scoped.actions.file:
                return scoped.actions.file == "quarantine"
        return self.store.has_action(target_type, name, "file", "quarantine")

    def block_for_connector(
        self, target_type: str, name: str, connector: str, reason: str,
    ) -> None:
        """Block ``name`` for ``connector`` (exact-match; connector="" = global)."""
        if self.store:
            self.store.set_action_field(
                target_type, name, "install", "block", reason, connector,
            )

    def allow_for_connector(
        self, target_type: str, name: str, connector: str, reason: str,
    ) -> None:
        """Allow ``name`` for ``connector`` and clear residual file/runtime state.

        Exact-match on connector (connector="" = global). Mirrors :meth:`allow`.
        """
        if not self.store:
            return
        self.store.set_action_field(
            target_type, name, "install", "allow", reason, connector,
        )
        self.store.clear_action_field(target_type, name, "file", connector)
        self.store.clear_action_field(target_type, name, "runtime", connector)

    def unblock_for_connector(
        self, target_type: str, name: str, connector: str = "",
    ) -> None:
        """Clear the install action for ``connector`` (exact-match; ""=global)."""
        if self.store:
            self.store.clear_action_field(target_type, name, "install", connector)

    def quarantine_for_connector(
        self, target_type: str, name: str, connector: str, reason: str,
    ) -> None:
        """Quarantine ``name`` for ``connector`` (file dimension; exact-match;
        connector="" = global). Mirrors :meth:`quarantine`."""
        if self.store:
            self.store.set_action_field(
                target_type, name, "file", "quarantine", reason, connector,
            )

    def clear_quarantine_for_connector(
        self, target_type: str, name: str, connector: str = "",
    ) -> None:
        """Clear the file (quarantine) action for ``connector`` (exact-match;
        ""=global). Mirrors :meth:`clear_quarantine`."""
        if self.store:
            self.store.clear_action_field(target_type, name, "file", connector)

    def disable_for_connector(
        self, target_type: str, name: str, connector: str, reason: str,
    ) -> None:
        """Disable ``name`` at runtime for ``connector`` (runtime dimension;
        exact-match; connector="" = global). Mirrors :meth:`disable`."""
        if self.store:
            self.store.set_action_field(
                target_type, name, "runtime", "disable", reason, connector,
            )

    def enable_for_connector(
        self, target_type: str, name: str, connector: str = "",
    ) -> None:
        """Clear the runtime (disable) action for ``connector`` (exact-match;
        ""=global). Mirrors :meth:`enable`."""
        if self.store:
            self.store.clear_action_field(target_type, name, "runtime", connector)

    def remove_action_for_connector(
        self, target_type: str, name: str, connector: str = "",
    ) -> None:
        """Remove all enforcement for ``connector`` (exact-match; ""=global)."""
        if self.store:
            self.store.remove_action(target_type, name, connector)

    # ------------------------------------------------------------------
    # Tool-level helpers (target_type="tool", scoped naming supported)
    # ------------------------------------------------------------------

    def is_tool_blocked(self, tool_name: str, source: str = "") -> bool:
        """Return True if the tool is blocked (scoped check first, then global)."""
        if not self.store:
            return False
        if source:
            scoped = f"{source}/{tool_name}"
            if self.store.has_action("tool", scoped, "install", "block"):
                return True
        return self.store.has_action("tool", tool_name, "install", "block")

    def is_tool_allowed(self, tool_name: str, source: str = "") -> bool:
        """Return True if the tool is allowed (scoped check first, then global)."""
        if not self.store:
            return False
        if source:
            scoped = f"{source}/{tool_name}"
            if self.store.has_action("tool", scoped, "install", "allow"):
                return True
        return self.store.has_action("tool", tool_name, "install", "allow")

    def block_tool(self, tool_name: str, source: str, reason: str) -> None:
        """Block a tool, optionally scoped to a source."""
        if self.store:
            target = f"{source}/{tool_name}" if source else tool_name
            self.store.set_action_field("tool", target, "install", "block", reason)

    def allow_tool(self, tool_name: str, source: str, reason: str) -> None:
        """Allow a tool, optionally scoped to a source.

        Uses the same cleanup pattern as allow() for consistency.
        """
        if not self.store:
            return
        target = f"{source}/{tool_name}" if source else tool_name
        self.store.set_action_field("tool", target, "install", "allow", reason)
        self.store.clear_action_field("tool", target, "file")
        self.store.clear_action_field("tool", target, "runtime")

    def list_blocked_tools(self) -> list[ActionEntry]:
        """List all tool-level block entries."""
        if not self.store:
            return []
        return self.store.list_by_action_and_type("install", "block", "tool")

    def list_allowed_tools(self) -> list[ActionEntry]:
        """List all tool-level allow entries."""
        if not self.store:
            return []
        return self.store.list_by_action_and_type("install", "allow", "tool")

    # ------------------------------------------------------------------
    # Connector-scoped tool helpers (target_type="tool", "@<connector>/<tool>")
    #
    # The ``@`` sigil keeps connector scoping distinct from the orthogonal
    # ``<source>/<tool>`` source scoping above. Runtime resolution order
    # (mirrored by the Go gateway lanes via the policy.go methods of the same
    # name) is, for request connector ``C`` and tool ``T``:
    #   block @C/T → allow @C/T → block T → allow T → scan
    # i.e. a connector-scoped row is authoritative for that connector; only
    # when no scoped action exists does the global row fall through.
    # ------------------------------------------------------------------

    @staticmethod
    def _tool_connector_target(tool_name: str, connector: str) -> str:
        """Build the connector-scoped tool key ``@<connector>/<tool>``.

        Centralised here so the read gate and the write surface stay in
        lockstep on the encoding.
        """
        return f"@{connector}/{tool_name}" if connector else tool_name

    def is_tool_blocked_for_connector(self, tool_name: str, connector: str = "") -> bool:
        """Return True if the tool is blocked for ``connector`` (scoped row, else global)."""
        if not self.store:
            return False
        if connector:
            scoped = self._tool_connector_target(tool_name, connector)
            entry = self.store.get_action("tool", scoped)
            if entry is not None and entry.actions.install:
                return entry.actions.install == "block"
        return self.store.has_action("tool", tool_name, "install", "block")

    def is_tool_allowed_for_connector(self, tool_name: str, connector: str = "") -> bool:
        """Return True if the tool is allowed for ``connector`` (scoped row, else global)."""
        if not self.store:
            return False
        if connector:
            scoped = self._tool_connector_target(tool_name, connector)
            entry = self.store.get_action("tool", scoped)
            if entry is not None and entry.actions.install:
                return entry.actions.install == "allow"
        return self.store.has_action("tool", tool_name, "install", "allow")

    def block_tool_for_connector(self, tool_name: str, connector: str, reason: str) -> None:
        """Block a tool, optionally scoped to a connector (``@<connector>/<tool>``)."""
        if self.store:
            target = self._tool_connector_target(tool_name, connector)
            self.store.set_action_field("tool", target, "install", "block", reason)

    def allow_tool_for_connector(self, tool_name: str, connector: str, reason: str) -> None:
        """Allow a tool, optionally scoped to a connector.

        Uses the same cleanup pattern as :meth:`allow_tool` for consistency.
        """
        if not self.store:
            return
        target = self._tool_connector_target(tool_name, connector)
        self.store.set_action_field("tool", target, "install", "allow", reason)
        self.store.clear_action_field("tool", target, "file")
        self.store.clear_action_field("tool", target, "runtime")

    def unblock_tool_for_connector(self, tool_name: str, connector: str = "") -> None:
        """Clear the install action for a connector-scoped tool row."""
        if self.store:
            target = self._tool_connector_target(tool_name, connector)
            self.store.clear_action_field("tool", target, "install")
